Sources/NIOTestUtils/NIOHTTP1TestServer.swift

```diff
@@ -0,0 +1,318 @@
+///                                                 method: .GET,
+///                                                 uri: "/some-route",
+///                                                 headers: .init([
+///                                                     ("Content-Type", "text/plain; charset=utf-8"),
+///                                                     ("Content-Length", "4")]))),
+///                                     try testServer.readInbound()))
+///     var requestBuffer = allocator.buffer(capacity: 128)
+///     requestBuffer.writeString(requestBody)
+///     XCTAssertNoThrow(XCTAssertEqual(.body(requestBuffer),
+///                                     try testServer.readInbound()))
+///     XCTAssertNoThrow(XCTAssertEqual(.end(nil),
+///                                     try testServer.readInbound()))
+///
+///     // Make the server send a response to the client.
+///     let responseBody = "pong"
+///     var responseBuffer = allocator.buffer(capacity: 128)
+///     responseBuffer.writeString(responseBody)
+///     XCTAssertNoThrow(try testServer.writeOutbound(.head(.init(version: .init(major: 1, minor: 1), status: .ok))))
+///     XCTAssertNoThrow(try testServer.writeOutbound(.body(.byteBuffer(responseBuffer))))
+///     XCTAssertNoThrow(try testServer.writeOutbound(.end(nil)))
+///
+///     // Assert that the client received the response from the server.
+///     XCTAssertNoThrow(XCTAssertEqual(responseBody, try requestComplete.wait()))
+public final class NIOHTTP1TestServer {
+    private let eventLoop: EventLoop
+    // all protected by eventLoop
+    private let inboundBuffer: BlockingQueue<HTTPServerRequestPart> = .init()
+    private var currentClientChannel: Channel? = nil
+    private var serverChannel: Channel! = nil
+
+    enum State {
+        case channelsAvailable(CircularBuffer<Channel>)
+        case waitingForChannel(EventLoopPromise<Void>)
+        case idle
+        case stopped
+    }
+    private var state: State = .idle
+
+    func handleChannels() {
+        self.eventLoop.assertInEventLoop()
+
+        let channel: Channel!
+        switch self.state {
+        case .channelsAvailable(var channels):
+            channel = channels.removeFirst()
+            if channels.isEmpty {
+                self.state = .idle
+            } else {
+                self.state = .channelsAvailable(channels)
+            }
+        case .idle:
+            let promise = self.eventLoop.makePromise(of: Void.self)
+            promise.futureResult.whenSuccess {
+                self.handleChannels()
+            }
+            self.state = .waitingForChannel(promise)
+            return
+        case .waitingForChannel:
+            preconditionFailure("illegal state \(self.state)")
+        case .stopped:
+            return
+        }
+
+        assert(self.currentClientChannel == nil)
+        self.currentClientChannel = channel
+        channel.closeFuture.whenSuccess {
```

Comment:
Do we need to consider what happens here if the connection is closed by the remote peer while we’re waiting to handle a previous connection? In this case I think we’d just lose the connection as this promise will fire immediately.